### ISSUE TITLE
Gerar Arquvivo Remessa CNAB400

### ISSUE THREAD

**Issue by SergioDeMoura** [BoletoNet/boletonet#97]
Olá pessoal,

como faço para gerar um aquivo remessa cnab400 para banco bradesco carteira 9?

o retorno resolvi mas nao estou conseguindo implementar a remessa.

obrigado

**Comment by carloscds** [comment]
@SergioDeMoura O código da remessa está aqui: https://github.com/BoletoNet/boletonet/blob/master/src/Boleto.Net/Banco/Banco_Bradesco.cs

**Comment by SergioDeMoura** [comment]
Oi, Carlos,

nao querendo ser preguiçoso mas por acaso teria algum exemplo implementado para geracao do arquivo remessa cnab400 para o banco bradesco assim como do itau? Pelo Itau consigo gerar mas, pelo bradesco nao estou conseguindo.

Muito obrigado!

Abraços!

**Comment by carloscds** [comment]
@SergioDeMoura no próprio código do BoletoNet tem vários exemplos.

**Comment by miloxsan** [comment]
ola Carlos. Estou tentando gerar arquivo retorno bradesco, e recebo o erro "A cadeia de caracteres de entrada não estava em um formato incorreto." Pode me ajudar ?

**Comment by carloscds** [comment]
@miloxsan Você verificou os dados da conta ? Em que local exatamente dá a mensagem ?

**Comment by miloxsan** [comment]
quando chamo esse metodo 

 arqRem.GerarArquivoRemessa("4837528", new Banco(237), cd, lb,
                arq, 01);

**Comment by carloscds** [comment]
@miloxsan vc validou os dados do banco ?

**Comment by miloxsan** [comment]
Hum, na geração do boleto sim usei o boletoValida() seria isso ou outra coisa ?

**Comment by carloscds** [comment]
@miloxsan os dados do seu convênio, que são fornecidos pelo banco

**Comment by miloxsan** [comment]
entao liguei la no bradesco eles me disseram q nao sabe com esse nome de cod convenio, perguntaram seria referente a qual coluna do arquivo seria ... vc sabe me dizer ?

**Comment by carloscds** [comment]
@miloxsan Cada banco trata a informação de uma maneira. alguns tem convênio, carteira, etc. Dê uma olhada no manual do banco e também qual o tipo de cobrança que a sua empresa está adotando.

**Comment by miloxsan** [comment]
Pois entao as informações que o banco nos passou foram: Ag Conta carteira 09 e codigo cedente. Pode me dizer oq causa essa falha na geração de detalhes do Arquivo Remessa ? Esse metodo GerarArquivoRemessa funciona dessa forma como passei ? E o que seria essa validação que vc disse ?

**Comment by carloscds** [comment]
@miloxsan Já deu uma olhada aqui: https://github.com/BoletoNet/boletonet/blob/ee5a36f8d7e388cdf88f54c795313fe4ee20a978/src/Boleto.Net.Testes/BancoBradescoTeste.cs

**Comment by miloxsan** [comment]
Sim, gerar o boleto consigo normalmente. Antes usavamos boleto sem registro entao era de boa ... agora q mudamos da carteira 06 para 09 e surgiu a necessidade do arquivo remessa. Esta dando erro de cadeia de caracteres "A cadeia de caracteres de entrada não estava em um formato incorreto." creio q este erro jamais seria causado por uma informação de codigo cedente, convenio etc errada correto ?

**Comment by carloscds** [comment]
@miloxsan Baixe o código do boletonet e faça um debug na sua aplicação junto com o codigo. Pode ser uma falha. Se for você pode consertar e submeter para o Git.

**Comment by miloxsan** [comment]
Blz vou ver esse metodo, mas sem querer abusar da sua boa vontate, vc tem um metodo ai que eu possa usar pra criar esse arquivo remessa passando uma lista de boletos ?

**Comment by carloscds** [comment]
@miloxsan Dá uma olhada nos exemplos que tem aqui no projeto.

**Comment by miloxsan** [comment]
Estao eu copiei os metodos TODOS rs e consegui estava dando erro no endereço sacado, mas qndo uso o metodo GerarArquivoRemessa continua dando erro de cadeia de caracteres. Gerou o arquivo eu enviei pro banco deu arquivo inválido. Pode me ajudar vou colar abaixo o arquivo remessa gerado.

01REMESSA01COBRANCA       00000000000004837528INTERMEIO SOLUCOES DE PAGAMENT237BRADESCO       231115        MX0000001                                                                                                                                                                                                                                                                                     000001
100000 000000000000 00090044900056840                         0000000009/1000000000000000002N           2  01000014798723111500000000010000000000099N23111500000000000000000000000000000000000000000000000000000000000000010001136399961LUIZ CARLOS                            RUA ANTONIO NATALIO RIGATIERI                      03635020                                                            000002
9                                                                                                                                                                                                                                                                                                                                                                                                         000003

**Comment by carloscds** [comment]
@miloxsan Você precisa verificar o manual do banco e conferir com o arquivo. Validando a posição que o banco diz que está errada.

**Comment by rafd75** [comment]
@miloxsan @carloscds --- cheguei nessa ISSUE... Acabei de depurar esse problema (está ocorrendo aqui também).

Vi uma série de problemas na criação da linha detalhe...

A rotina boleto.Valida formata o nosso número assim: 09/12345678901-2 e ao gerar a remessa, por estar validado dessa forma, fica todo errado:

Problema 1:
No método GerarDetalheRemessaCNAB400, na linha abaixo, ao invés de retornar 12345678901 (que é o nosso número), está retornando "09/12345678" para _detalhe.

//Identificação do Título no Banco (12, A)
_detalhe += Utils.FitStringLength(boleto.NossoNumero, 11, 11, '0', 0, true, true, true); //Nosso Numero (11)

Logo em seguida (código abaixo), tenta adicionar o dígito do nosso numero, mas observe que na chamada do método Mod11Bradesco, é passado como parametro a carteira "boleto.Carteira" + boleto.NossoNumero (que já contém a carteira, lembra?) 

// Força o NossoNumero a ter 11 dígitos. Alterado por Luiz Ponce 07/07/2012
_detalhe += Mod11Bradesco(boleto.Carteira + Utils.FitStringLength(boleto.NossoNumero, 11, 11, '0', 0, true, true, true), 7); // Digito de Auto Conferencia do Nosso Número (01)

Ou seja, Mod11Bradesco recebe algo assim: 0909/12345678
Provavelmente esse erro "Cadeia de caracteres" é por causa da barra que está indo para o cálculo do Mod11Bradesco.

----- 

Sou novato nesse projeto, sou novato no C#, mas sou macaco velho de lógica de programação e acho que esse código "Remessa" do Bradesco (Cnab400) não funcionou nunca.

Minha dúvida é se eu posso corrigir isso? Se realmente não funciona, tem que ser corrigido... mas se alguém estiver usando isso dessa forma (talvez preenchendo as propriedades do boleto de forma diferente), vai parar tudo.... 

Aguardo retorno!

**Comment by carloscds** [comment]
@rafd75 @miloxsan Podem corrigir a rotina, pois parece que está realmente com problema.

**Comment by rafd75** [comment]
Antes de alterar a rotina do Bradesco, fiz um teste com a rotina do Itaú. O erro "Cadeia de caracteres" não ocorre, e o arquivo .REM é gerado. Mas dentro do arquivo .REM está com a / que separa a carteira do nosso número... vou continuar estudando aqui.

**Comment by rafd75** [comment]
Talvez eu matei o problema aqui.

No caso de imprimir os boletos:

boleto.Valida();
boletos.Add(boleto);

No caso de gerar remessa:
boletos.Add(boleto);

Dessa forma, nem no Itau, nem no Bradesco, o problema ocorre.

**Comment by rafaelreisathenas** [comment]
Bom dia rafd75, 

estou passando pelo mesmo problema de ter que implementar a remessa Bradesco para a carteira 9 e pelo que ví no código do boleto.net (projeto Arquivo), esta remessa para o banco 237 não foi implementada.

Você conseguiu resolver seu problema? Poderia me passar alguma dica para gerar tal remessa?

**Comment by rafd75** [comment]
@rafaelreisathenas - Bradesco Carteira 9 funcionando aqui.

Obs: O problema que eu tive, comentei na mensagem acima... Se for para imprimir o boleto, chamar o Valida() antes de adicionar o boleto na coleção boletos.... Se for para gerar a remessa., não chamar o Valida() - use apenas o Add(boleto).

**Comment by rafd75** [comment]
respondi pelo github...

vc tem um código de exemplo para eu ver? está recebendo alguma msg de erro ?




*Rodolfo Duarte**Computar* | Desenvolvimento
(16) 3011-2768 / 3043-9460
rodolfo@computar.com.br


[image: Computar] <http://www.computar.com.br>









Em 19 de janeiro de 2016 11:42, rafaelreisathenas <notifications@github.com>
escreveu:

**Comment by vulock** [comment]
Galera eu estou usando o bradesco e realmente nao funciona
alguns eros que estou debugando:
1 no valor do boleto, ele está dando o replace pela virgula, nao pelo ponto, e no bradesco nao pode ir isso, fiz mais um replace para o ponto (alem da virgula)
2 mesma coisa pro campo de juros
3 nao esquecer da linha 
b.EspecieDocumento = new EspecieDocumento_Bradesco("1");
se nao ele da erro de null no arquivo (pode ser o que estava dando para voces ai em cima)

quando eu acabar eu posto o resultado

**Comment by vulock** [comment]
Exemplo gerando 2 boletos do bradesco e a respectiva remessa usando o exemplo do projeto
main.cs
`public void GeraDadosBradesco()
        {
            DateTime vencimento = new DateTime(2007, 9, 10);

            Instrucao_Bradesco item1 = new Instrucao_Bradesco(9);
            Instrucao_Bradesco item2 = new Instrucao_Bradesco(9);
            Cedente c = new Cedente("00.000.000/0000-00", "Empresa de Atacado", "1229", "7","11111","8");
            //Na carteira 198 o código do Cedente é a conta bancária
            c.Codigo = "13000";

            Boleto b = new Boleto(vencimento, 1642, "9", "92082835",c);
            b.NumeroDocumento = "1008073";

            b.DataVencimento = Convert.ToDateTime("12-12-12");
            b.EspecieDocumento = new EspecieDocumento_Bradesco("1");
            
            b.Sacado = new Sacado("000.000.000-00", "Fulano de Silva");
            b.Sacado.Endereco.End = "SSS 154 Bloco J Casa 23";
            b.Sacado.Endereco.Bairro = "Testando";
            b.Sacado.Endereco.Cidade = "Testelândia";
            b.Sacado.Endereco.CEP = "70000000";
            b.Sacado.Endereco.UF = "DF";

            item2.Descricao += item2.QuantidadeDias.ToString() + " dias corridos do vencimento.";
            b.Instrucoes.Add(item1);
            b.Instrucoes.Add(item2);
            b.Cedente.ContaBancaria.DigitoAgencia = "1";
            b.Cedente.ContaBancaria.DigitoAgencia = "2";

            b.Banco = new Banco(237);

            Boletos boletos = new Boletos();
            boletos.Add(b);

            Boleto b2 = new Boleto(vencimento,5900m, "9", "92082835", c);
            b2.NumeroDocumento = "1008073";

            b2.DataVencimento = Convert.ToDateTime("12-12-12");
            b2.EspecieDocumento = new EspecieDocumento_Bradesco("1");
     
            b2.Sacado = new Sacado("000.000.000-00", "Fulano de Silva");
            b2.Sacado.Endereco.End = "SSS 154 Bloco J Casa 23";
            b2.Sacado.Endereco.Bairro = "Testando";
            b2.Sacado.Endereco.Cidade = "Testelândia";
            b2.Sacado.Endereco.CEP = "70000000";
            b2.Sacado.Endereco.UF = "DF";

            item2.Descricao += item2.QuantidadeDias.ToString() + " dias corridos do vencimento.";
            b2.Instrucoes.Add(item1);
            b2.Instrucoes.Add(item2);
            b2.Cedente.ContaBancaria.DigitoAgencia = "1";
            b2.Cedente.ContaBancaria.DigitoAgencia = "2";

            b2.Banco = new Banco(237);

            boletos.Add(b2);

            GeraArquivoCNAB400(b2.Banco, c, boletos);
        }`
no Banco_Bradesco.cs
onde tiver referenciando valores
`_detalhe += Utils.FitStringLength(boleto.JurosMora.ToString("0.00").Replace(",", ""), 13, 13, '0', 0, true, true, true)`

adicione o `.Replace(",","")`
assim ele substituira virgulas E pontos por ""
#66 
@carloscds

**Comment by carisoftware** [comment]
Consertaram esse erro no projeto?

**Comment by Quallitec** [comment]
Como faço para gerar varios registros (boletos) em um arquivo de remessa apenas.

Estou usando o banco bradesco.

**Comment by six2016** [comment]
Olá muito bom dia, gostaria de uma ajuda de vocês , estou a duas semanas tentando gerar arquivo de remessa até ai tudo ok, só que no arquivo retorno do banco chega o seguinte erro: ERROS ENCONTRADOS: REG.TIPO 1 - COD.DE ERRO POSICOES 30/31 E 74/75.    Alguém poderia me dizer o que pode ser ,
conferi os dados da empresa e estão corretos.

**Comment by vulock** [comment]
qual banco!? tente ligar no suporte do banco, pelo menos o bradesco me
ajudou muito.

Em 13 de dezembro de 2016 09:10, six2016 <notifications@github.com>
escreveu:

**Comment by six2016** [comment]
Ola  Vulock muito obrigado.
O banco é o Caixa, acho que consegui resolver o arquivo retorno já não retornou erro, entrarei em contacto com o Cliente e verificar mas obrigado.

**Comment by joelcc** [comment]
Bom dia @carloscds 

Estou com o mesmo problema do pessoal ao tentar gerar o arquivo remessa para o Bradesco na carteira 09 a msg de erro que me aparece é a seguinte:

System.Exception: Erro durante a geração dos registros de DETALHE do arquivo de REMESSA. ---> System.Exception: Erro durante a geração do DETALHE arquivo de REMESSA. ---> System.Exception: Erro ao gerar DETALHE do arquivo CNAB400. ---> System.Exception: Problemas ao Formatar a string. String = ---> System.NullReferenceException: Referência de objeto não definida para uma instância de um objeto. em BoletoNet.Utils.FitStringLength(String SringToBeFit, Int32 maxLength, Int32 minLength, Char FitChar, Int32 maxStartPosition, Boolean maxTest, Boolean minTest, Boolean isNumber) --- Fim do rastreamento de pilha de exceções internas --- em BoletoNet.Utils.FitStringLength(String SringToBeFit, Int32 maxLength, Int32 minLength, Char FitChar, Int32 maxStartPosition, Boolean maxTest, Boolean minTest, Boolean isNumber) em BoletoNet.Banco_Bradesco.GerarDetalheRemessaCNAB400(Boleto boleto, Int32 numeroRegistro, TipoArquivo tipoArquivo) --- Fim do rastreamento de pilha de exceções internas --- em BoletoNet.Banco_Bradesco.GerarDetalheRemessaCNAB400(Boleto boleto, Int32 numeroRegistro, TipoArquivo tipoArquivo) em BoletoNet.Banco_Bradesco.GerarDetalheRemessa(Boleto boleto, Int32 numeroRegistro, TipoArquivo tipoArquivo) --- Fim do rastreamento de pilha de exceções internas --- em BoletoNet.Banco_Bradesco.GerarDetalheRemessa(Boleto boleto, Int32 numeroRegistro, TipoArquivo tipoArquivo) em BoletoNet.Banco.GerarDetalheRemessa(Boleto boleto, Int32 numeroRegistro, TipoArquivo tipoArquivo) --- Fim do rastreamento de pilha de exceções internas --- em BoletoNet.Banco.GerarDetalheRemessa(Boleto boleto, Int32 numeroRegistro, TipoArquivo tipoArquivo) em BoletoNet.ArquivoRemessaCNAB400.GerarArquivoRemessa(String numeroConvenio, IBanco banco, Cedente cedente, Boletos boletos, Stream arquivo, Int32 numeroArquivoRemessa)

**Comment by vulock** [comment]
Bom dia!
ele está reclamando que está faltando algum dado, debuga pra ver quem está
faltando.

Em 9 de janeiro de 2017 13:11, joelcc <notifications@github.com> escreveu:

**Comment by joelcc** [comment]
Boa trade @vulock !
Tem como vc me auxiliar no debug? Pois a versão que uso aqui do Arquivo não gerar remessa do Bradesco para nenhum dos 2 tipos.

**Comment by vulock** [comment]
voce fez o download do projeto ou so da dll? eu tive que fazer algumas
alteracoes no projeto

Em 9 de janeiro de 2017 15:43, joelcc <notifications@github.com> escreveu:

**Comment by diegopgcabral** [comment]
Boa tarde!
Sou novato no desenvolvimento em C# e estou precisando implementar a geração de boleto e remessa para o Bradesco, porém não estou conseguindo....já consegui para o BB e Itau....Porem para o Bradesco sempre recebo critica....Como se a carteira do Bradesco não tivesse sido implementada.
Consegui um exemplo do Itau para gerar, mas mudando para o Bradesco sempre recebo o erro que não está implementada para esse Banco...
Alguém consegue me dar uma luz?

**Comment by joelcc** [comment]
Fiz a instalação usando o NuGet, por favor então ajuda ai. Manda o link que vou baixar, desde já agradecido.

**Comment by matuskelias** [comment]
Bom dia @carloscds,
Consegui arrumar o erro? Se conseguiu, posta aqui o que você fez.

**Comment by TelesGabriiel** [comment]
Esse problema ocorre por que na hora de passar o nosso numero esta passando com caracteres especias tentem da um Regex.Replace(Boleto.NossoNumero, "[^0-9a-zA-Z]+", ""); antes de chamar o metodo GerarArquivoRemessa();

**Comment by vulock** [comment]
é só colocar o ponto "." na lista, dependendo do sistema e da linguagem,
ele coloca ponto e nao virgula para decimais.

2017-05-30 0:11 GMT-03:00 TelesGabriiel <notifications@github.com>:

**Comment by fellpzns** [comment]
Bom dia, após baixar o projeto e instalar o NuGet, preciso fazer mais alguma coisa? Estou tentando debugar porém ele diz que não é possível e critica em 63 erros. :(

**Comment by eternoscav** [comment]
Boa tarde pessoal acordando novamente o tópico,
alguém poderia me dizer se a dll do boletonet para geração do arquivo de remessa do bradesco está funcionando no meu da "Erro ao gerar arquivo remessa" erro 610.

Quem poderia me ajudar?
Obrigado.

**Comment by vulock** [comment]
Eu uso, funciona. O que eu tive que ajustar eh a cultura para deletar os
pontos e virgulas.

Em sex, 20 de jul de 2018 17:50, eternoscav <notifications@github.com>
escreveu:

**Comment by vulock** [comment]
Sim, no meu ambiente as maquinas sao todas em inglês, logo o dogoto
separador dr decimal é o ponto, nao a virgula...
Ele chega a gerar o arquivo ou nao? Tem q ver em qual linha que para

Em sáb, 21 de jul de 2018 08:27, eternoscav <notifications@github.com>
escreveu:

**Comment by eternoscav** [comment]
Entendi, então no meu caso ele para nessa linha "                        objRemessa.GerarArquivoRemessa(numConvenio, New Banco(CodBanco), c, boletos, arq, 1)
"

**Comment by vulock** [comment]
Vc ta usando a dll direto? Compila o projeto e ve o q ta dando, é melhor.

Em sáb, 21 de jul de 2018 08:31, eternoscav <notifications@github.com>
escreveu:

**Comment by eternoscav** [comment]
Eu uso a Dll da data 14/06/2018 e hoje tenho funcionando o Caixa, Sicoob, Sicred tds funcionam boleto e Remessa,
quando compilo não me apresenta erro algum, é só quando vou executar o método, que me trás o erro.

**Comment by vulock** [comment]
Sim, pode ser alguma especificacao do bradesco que esta faltando por
exemplo. Eu sódesenvolvi pro bradesco e foi normal, uso a dll de uns 2 anos
atras.

Em sáb, 21 de jul de 2018 08:42, eternoscav <notifications@github.com>
escreveu:

**Comment by eternoscav** [comment]
Em qual linguagem está o seu projeto?
Poderia me passar um exemplo, pois os que consegui tentei tds porém sem sucesso.

**Comment by vulock** [comment]
C#, pode me lembrar na segunda? Nao estou no escritório hj.

Em sáb, 21 de jul de 2018 08:49, eternoscav <notifications@github.com>
escreveu:

**Comment by eternoscav** [comment]
Sim, claro é quase certeza que ele esta pedindo algum parâmetro que eu não estou passando, o problema é descobrir que parâmetro ou até mesmo algum que está sendo passado errado.
Mas segunda eu volto a te incomodar.

Muito Obrigado pela atenção.

**Comment by eternoscav** [comment]
Á esse é o código que estou usando:
```
 Dim codigo As String = Cedente_Codigo.Replace("-", "").Replace(" ", "")
                    c = New Cedente("00000000000000", "Empresa de Atacado", "1234", "5", "123456", "7")
                    c.Codigo = 50 'VERIFICAR

                    b = New Boleto(vencimento, Valor, "09", "01000000001", c)
                    b.NumeroDocumento = "01000000001"
                    b.DataDocumento = Now

                    b.Sacado = New Sacado("00000000000", "Nome do seu Cliente")
                    b.Sacado.Endereco.End = "Endereço do seu Cliente"
                    b.Sacado.Endereco.Bairro = Sacado_Bairro
                    b.Sacado.Endereco.Cidade = Sacado_Cidade
                    b.Sacado.Endereco.CEP = Sacado_CEP
                    b.Sacado.Endereco.UF = Sacado_UF

                    Dim item As Instrucao_Bradesco = New Instrucao_Bradesco()
                    item.Descricao = dados_para_instrucao
                    b.Instrucoes.Add(item)
                    b.Aceite = aceite
                    boletoBancario.AjustaTamanhoFonte(10, 10, 10, 10)
                    boletoBancario.FormatoCarne = carne
                    boletoBancario.GerarArquivoRemessa = True
                    boletoBancario.RemoveSimboloMoedaValorDocumento = True

                    b.Remessa = New Remessa()
                    b.Remessa.TipoDocumento = "2"

                    boletoBancario.Boleto = b
                    boletoBancario.Boleto.Valida()

                    Dim objRemessa As New ArquivoRemessa(TipoArquivo.CNAB400)
                    Dim arq As System.IO.FileStream = New System.IO.FileStream(nome_arq_remessa, System.IO.FileMode.Append)
                        Dim boletos As Boletos = New Boletos()
                        Dim strMensagem As String = ""

                        boletos.Add(b)

                        numArqRemessa = Convert.ToInt32(contRem)
                        objRemessa.GerarArquivoRemessa("09", New Banco(237), c, boletos, arq, numArqRemessa)
```

**Comment by eternoscav** [comment]
Bom dia!
Não esqueça de mim, quando puder mandar o exemplo do código com os parâmetros para a geração da remessa.

Obrigado.

**Comment by eternoscav** [comment]
Ok, vou estudar o código aqui e ver o que estou fazendo errado, daqui a pouco com mais informações mas enquanto muito obrigado.

**Comment by vulock** [comment]
qualquer duvida estamos aí

Em seg, 23 de jul de 2018 às 09:28, eternoscav <notifications@github.com>
escreveu:

**Comment by pcmafei** [comment]
Bom dia, na verdade, se tentar gerar arquivo remessa somente para um boleto esta ok.

O problema se da quando tentamos gerar srquivo de remessa com varios boletos

Obter o Outlook para Android<https://aka.ms/ghei36>

________________________________

**Comment by eternoscav** [comment]
Boa noite Pessoal!

Tenho o banco Bradesco funcionando tanto boleto como Arquivo de Remessa,
mas precisei colocar algumas instruções e incluindo o sistema não consegue gerar,
alguém poderia me ajudar?Alguma sugestão.

Quando incluo no meu projeto o 
"'Dim item As Instrucoes = New Instrucao_Bradesco(9,5)"
"b.Instrucoes.Add(item)"

Ele dá erro de Geração de arquivo de Remessa"

Obrigado.

**Comment by CompusoftsInformatica** [comment]
Boa tarde, estou com um problema ao gerar arquivo de remessa no gcloud, no meu ambiente funciona normal, mas no ambiente do google cloud (php  7 linux), quando gero o arquivo a ultima linha onde tem que ficar o ponteiro fica escrito null
[00286724 (1).txt](https://github.com/BoletoNet/boletonet/files/4974004/00286724.1.txt)
